src/nrnpython/nrnpy_nrn.cpp

```diff
@@ -269,6 +269,7 @@ static NPyMechObj* new_pymechobj(NPySegObj* pyseg, Prop* p) {
     m->pyseg_ = pyseg;
     Py_INCREF(m->pyseg_);
     m->prop_ = p;
+    new (&m->prop_id_) neuron::container::non_owning_identifier_without_container;
```

Comment:
> allocates m->prop_id_ such that its memory contains unknown (possibly non zeroed) data

I think this is exactly the issue. The reason it would work is that when it's zeroed out, the shared pointer will see a `nullptr` and therefore not attempt to free whatever the internal pointer points to. However, when its fillled with non-zeros, it doesn't know that what it's holding isn't a valid address. Hence, the assignment operator, will dereference internal pointers, either the control-block (which contains the reference count, or the pointer itself to free the object. Neither is particularly likely to point to an address mapped to our process. Hence, the segfault.

What placement new does is the C++ equivalent of a `memset` after a `malloc`, but instead of setting the memory to all `zeros` it calls the default constructor and put the default constructed object at the specified memory location `&m->prop_id_`.